commands/find_stemcell_version.go

```diff
@@ -0,0 +1,138 @@
+package commands
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/Masterminds/semver"
+	"github.com/pivotal-cf/jhanda"
+	"github.com/pivotal-cf/kiln/fetcher"
+	"github.com/pivotal-cf/kiln/internal/cargo"
+	"gopkg.in/src-d/go-billy.v4/osfs"
+	"log"
+	"regexp"
+	"strings"
+)
+
+const (
+	ErrStemcellOSInfoMustBeValid       = "Stemcell OS information is missing or invalid"
+	ErrStemcellMajorVersionMustBeValid = "Stemcell Major Version is missing or invalid"
+	TanzunetRemotePath                  = "network.pivotal.io"
+)
+
+type FindStemcellVersion struct {
+	outLogger   *log.Logger
+	pivnetService *fetcher.Pivnet
+
+	Options struct {
+		Kilnfile         string  `short:"kf" long:"kilnfile" default:"Kilnfile" description:"path to Kilnfile"`
+		VariablesFiles []string `short:"vf" long:"variables-file" description:"path to variables file"`
+		Variables      []string `short:"vr" long:"variable" description:"variable in key=value format"`
+	}
+}
+
+type stemcellVersionOutput struct {
+	Version    string `json:"version"`
+	Source     string `json:"source"`
+	RemotePath string `json:"remote_path"`
+}
+
+func NewFindStemcellVersion(outLogger *log.Logger, pivnetService *fetcher.Pivnet) FindStemcellVersion {
+	return FindStemcellVersion{
+		outLogger:   outLogger,
+		pivnetService: pivnetService,
+	}
+}
+
+func (cmd FindStemcellVersion) Execute(args []string) error {
+	kilnfile, err := cmd.setup(args)
+
+	if err != nil {
+		return err
+	}
+
+	var productSlug=""
+
+	//Get Stemcell OS and major from Kilnfile
+	if kilnfile.Stemcell.OS == "ubuntu-xenial" {
+		productSlug = "stemcells-ubuntu-xenial"
+	} else if kilnfile.Stemcell.OS == "windows2019" {
+		productSlug = "stemcells-windows-server"
+	}
+
+    if productSlug == "" {
+		return fmt.Errorf(ErrStemcellOSInfoMustBeValid)
+	} else if kilnfile.Stemcell.Version == "" {
+		return fmt.Errorf(ErrStemcellMajorVersionMustBeValid)
+	} else {
+
+		majorVersion, err := ExtractMajorVersion(kilnfile.Stemcell.Version)
+
+		if err != nil {
+			return err
+		}
+
+		//Get stemcell version from pivnet
+		stemcellVersion, err := cmd.pivnetService.StemcellVersion(productSlug, majorVersion)
+
+		if err != nil {
+			return err
+		}
+
+		stemcellVersionJson, _ := json.Marshal(stemcellVersionOutput{
```

Comment:
I would at least return the marshal error. It is rare that something would go wrong but ignoring errors should be a rare edge case. I would not write a test for such an error check.

Reply:
This one makes sense. I am returning the error now.